text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,748 @@
+    }
+}
+```
+
+This is a SemVer-breaking change. If the library author goes though with this, the aforementioned
+downstream code will no longer compile. In this scenario, the library author cannot use `unsafe` to
+denote that this field carries a safety invariant; this is *both* a hard violation of [**Tenet:
+Unsafe Fields Denote Safety Invariants**](#tenet-unsafe-fields-denote-safety-invariants), and (in
+requiring trivially `unsafe` drop glue), a violation of [**Tenet: Safe Usage is Usually
+Safe**](#tenet-safe-usage-is-usually-safe).
+
+# Drawbacks
+
+## Alarm Fatigue
+
+Although the `unsafe` keyword gives Rust's safety hygiene tooling insight into whether a field
+carries safety invariants, it does not give Rust deeper insight into the semantics of those
+invariants. Consequently, Rust must err on the side caution, requiring `unsafe` for most uses of
+unsafe field — including uses that the programmer can see are conceptually harmless.
+
+In these cases, Rust's safety hygiene tooling will suggest that the harmless operation is wrapped
+in an `unsafe` block, and the programmer will either:
+
+- comply and provide a trivial safety proof, or
+- opt out of Rust's field safety tooling by removing the `unsafe` modifier from their field.
+
+The former is a private annoyance; the latter is a rebuttal of Rust's safety hygiene conventions
+and tooling. Such rebuttals are not unprecedented in the Rust ecosystem. Even among prominent
+projects, it is not rare to find a conceptually unsafe function that is not marked unsafe. The
+discovery of such functions by the broader Rust community has, occasionally, provoked controversy.
+
+This RFC takes care not to fuel such flames; e.g., [**Tenet: Unsafe Fields Denote Safety
+Invariants**](#tenet-unsafe-fields-denote-safety-invariants) admonishes that programmers *should* —
+but **not** *must* — denote field safety invariants with the `unsafe` keyword. It is neither a
+soundness nor security issue to continue to adhere to the convention of using visibility to enforce
+field safety invariants.
+
+This RFC does not, itself, attempt to address alarm fatigue. Instead, we propose a simple extension
+to Rust's safety tooling that is, by virtue of its simplicity, particularly amenable to future
+iterative refinement. We imagine empowering Rust to reason about safety invariants, either via
+[type-directed analyses](#safe-reads-for-fields-with-local-non-suspended-invariants), or via
+syntactic extensions. The design of these refinements will be guided by the valuable usage data
+produced by implementing this RFC.
+
+# Prior art
+
+Some items in the Rust standard library have `#[rustc_layout_scalar_valid_range_start]`,
+`#[rustc_layout_scalar_valid_range_end]`, or both. These items have identical behavior to that of
+unsafe fields described here. It is likely (though not required by this RFC) that these items will
+be required to use unsafe fields, which would reduce special-casing of the standard library.
```

Comment:
Aren't these attributes also used to guide niche optimization? IIUC, a `NonZeroU8` defined merely using an `unsafe` field would not inform Rust that `0` is an invalid value, and so wouldn't be subject to existing niche optimizations.

Reply:
I read this language as saying that unsafe fields would be required in addition to the attributes.

Reply:
The reason for that language is not about niche value optimization, but rather the requirement of using `unsafe` to initialize or mutate it. If it's being removed, so be it, but it's a good point of reference as to how things work at the moment.